Add dist language support

mferrera opened the pull request:
This PR adds support for the dist language and includes a mini-library for the probability distributions used by ERT.

Closes #35 
Closes #42

rubenthoms reviewed (changes requested):
LGTM, great job :+1: 
That was a lot of code but good to read and understand (I didn't go to deep in the math functions though 😛 ). A couple of minor comments to consider.

mferrera commented on src/languages/dist/providers/keywords-completion-provider.ts:
```diff
@@ -0,0 +1,31 @@
+            const word = model.getWordUntilPosition(position);
+            if (!options.completion) {
+                return { suggestions: [] };
+            }
```
Good catch! Fixed

mferrera commented on src/languages/dist/utils.ts:
```diff
@@ -0,0 +1,202 @@
+        const reader = new FileReader();
+        reader.onload = () => resolve(reader.result as string);
+        reader.onerror = () => reject(reader.error);
+        reader.onabort = () => reject(new Error("Read aborted"));
```
I gave more meaningful messages. Given the nature and repeatability of when this function is triggered with hovering it might not be suitable to bring attention to users. Hopefully error markers can cover that in a more meaningful way eventually. But agreed about thinking toward an error dispatcher of some sort.

mferrera commented:
Addressed your comments and prettified as well. Thanks a lot for taking on this much-too-large review! I will break them up into smaller ones next time

rubenthoms commented on src/languages/dist/utils.ts:
```diff
@@ -0,0 +1,197 @@
+    svgYAxis.selectAll("text").style("stroke", "gray");
+};
+
+const appendLine = (
```
Add return types?

rubenthoms commented on src/languages/dist/utils.ts:
```diff
@@ -0,0 +1,197 @@
+
+const PLOT_MARGINS = {top: 10, right: 10, bottom: 40, left: 40};
+
+const appendScaleX = (xVals: number[], size: Size) => {
```
Not sure about the naming. As I understand the function it takes an array of scalars and scales them according to `size`. How is it appending anything?

rubenthoms commented on src/languages/dist/utils.ts:
```diff
@@ -0,0 +1,197 @@
+
+import {modePalette} from "../../themes/theme";
+
+export const tokenizeLine = (model: editor.IReadOnlyModel, lineNumber: number): IDistToken[] => {
```
A general question: It might be better (for debugging) to use `function` instead of arrow functions. Arrow functions become anonymous in the debugger stack and are therefore harder to read/understand.  I know that I've also used a bunch of arrow functions, but we could decide to stick with functions from now (whenever it makes sense) and refactor the old arrow functions whenever it's convenient.

Here is a good article about the usage of arrow functions: https://zendev.com/2018/10/01/javascript-arrow-functions-how-why-when.html

mferrera replied:
Sounds fair to me. I adopted this as a more functional style and after running against callback difficulties with passing distribution pdf functions, but I guess this is the kind of nuance you pick up with more time using a language :)

rubenthoms commented on src/utils/distributions.ts:
```diff
@@ -0,0 +1,681 @@
+
+    inv = (p: number): number => {
+        if (this._min && p) {
+            throw new Error("Not implemeneted!");
```
Typo :wink:

mferrera replied:
Thanks! Just a comment on this PERT distribution while we're on it: it's not currently used at all or supported in .dist files, but it is supported in design matrices. I wanted to sneak the bulk of it in now during my detour to math land, but it will need a quick touch-up eventually.

rubenthoms reviewed (changes requested):
Nice fixes :+1: Again some minor comments only.

rubenthoms reviewed (approved):
LGTM!
